numbat/tests/interpreter.rs

```diff
@@ -817,3 +834,42 @@ fn test_statement_pretty_printing() {
     // TODO:
     // expect_pretty_print("fn f<Z>(z: Z) = z", "fn f<Z>(z: Z) -> Z = z");
 }
+
+#[test]
+fn test_assert_eq_3() {
+    // approximate equality succeeds when epsilon is larger than the difference between lhs and rhs
+    insta::assert_snapshot!(succeed("assert_eq(200cm, 201cm, 2cm)"), @"");
+
+    insta::assert_snapshot!(fail("assert_eq(200cm, 201cm, 2 watts)"), @"Argument types in assert_eq calls must match");
+
+    // approximate equality succeeds when epsilon is equal to the difference between the lhs and rhs
+    insta::assert_snapshot!(succeed("assert_eq(200cm, 201cm, 1cm)"), @"");
+
+    // approximate equality failure with the same units for all arguments shows no original values in parentheses
+    insta::assert_snapshot!(fail("assert_eq(200cm, 201cm, 0.1cm)"), @r###"
+    Assertion failed because the following two quantities differ by 1 cm, which is more than 0.1 cm:
+      200 cm
+      201 cm
+    "###);
+
+    // approximate equality failure with different units for rhs shows original values in parentheses for rhs only
+    insta::assert_snapshot!(fail("assert_eq(2000mm, 201cm, 0.1cm to mm)"), @r###"
+    Assertion failed because the following two quantities differ by 10.0 mm, which is more than 1 mm:
+      2000 mm
+      2010.0 mm (201 cm)
```

Comment:
It feels like this could be improved even further by choosing the same number of decimal digits for both values. But maybe that's kind of hard to do properly. So I'm also fine with merging this as is.

Reply:
I think it would be better to address this in a new issue/PR as it could be quite complex and may be relevant to and affect other places where pretty printing of numbers is done.

There is some floating point math imprecision which causes the conversion of 201 cm to mm to output `2010.0000000000002`.

This can be seen from a small test like so:
```rust
#[test]
fn conversion_floating_point() {
    let length = Quantity::new_f64(201.0, Unit::centimeter());
    let length = length.convert_to(&Unit::millimeter()).unwrap();

    assert_eq!(length.unsafe_value().to_f64(), 2010.0);
}
```

The `pretty_dtoa` crate formats the floating point value with 6 significant digits in `numbat/src/number.rs::pretty_print`. For `2010.0000000000002`, this should give `2010.00` or so, but additional code trims extra 0s to a single 0.

For a number like `212121001.123456`, it would return `212121000` or `2.12121e+8` in scientific notation, and we may want to change this for clarity depending on epsilon. We may need to account for how many decimal places lhs, rhs, epsilon and maybe delta would have if pretty printed and use the maximum.

I'm not sure how realistic this example is, but you can see here that the pretty printing hides the difference between the lhs and rhs.
```rust
insta::assert_snapshot!(fail("assert_eq(212121000.123456cm, 212121001.123456cm, 0.9cm)"), @r###"
Assertion failed because the following two quantities differ by 1 cm, which is more than 0.9 cm:
  2.12121e+8 cm
  2.12121e+8 cm
"###);
```

Separating these issues would avoid this complexity and prevent it from slowing down the improvement offered here.

Reply:
Perhaps I was overthinking it a bit though. We could probably just format all the values to use the number of decimal places in epsilon to keep it simple. There may be some edge cases where epsilon may have more decimal places than expected due to unit conversion or the use of an expression that we can ignore until a more precise underlying type is used for Quantities.

Example:
```rust
insta::assert_snapshot!(fail("assert_eq(2.0000006cm, 243.311336cm, 0.0001mm)"), @r###"
Assertion failed because the following two quantities differ by 2413.1134 mm, which is more than 0.0001 mm:
  0020.0000 mm (2.0 cm)
  2433.1134 mm (243.311 cm)
"###);
```

Reply:
> We could probably just format all the values to use the number of decimal places in epsilon to keep it simple.

That sounds very reasonable to me (I hope there is a clean way to extract the number of necessary digits from the value of epsilon). Maybe we can just add one digit as a safety margin (unless we want to think this through in more detail and figure out it's not necessary… or that even more digits might be necessary).